AU2/plugins/custom_plugins/PolicePlugin.py

```diff
@@ -0,0 +1,228 @@
+from AU2.html_components.InputWithDropDown import InputWithDropDown
+from AU2.html_components.IntegerEntry import IntegerEntry
+from AU2.html_components.Label import Label
+from AU2.html_components.NamedSmallTextbox import NamedSmallTextbox
+from AU2.plugins.AbstractPlugin import AbstractPlugin, Export, ConfigExport
+from AU2.plugins.CorePlugin import registered_plugin
+from AU2.plugins.constants import COLLEGES, WATER_STATUSES, WEBPAGE_WRITE_LOCATION
+from AU2.plugins.util import random_data
+from AU2.plugins.util.DeathManager import DeathManager
+from AU2.plugins.util.PoliceRankManager import DEFAULT_RANKS, PoliceRankManager
+
+POLICE_TABLE_TEMPLATE = """
+<p xmlns="">
+      Here is a list of members of our loyal police, charged with protecting Cambridge from murderers of innocents and people that annoy the all powerful Umpire:
+   </p>
+<table xmlns="" class="playerlist">
+  <tr><th>Rank</th><th>Pseudonym</th><th>Real Name</th><th>Email Address</th><th>College</th><th>Notes</th></tr>
+  {ROWS}
+</table>
+"""
+
+POLICE_TABLE_ROW_TEMPLATE = """
+<tr><td>{RANK}</td><td>{PSEUDONYM}</td><td>{NAME}</td><td>{EMAIL}</td><td>{COLLEGE}</td><td>{NOTES}</td></tr>
+"""
+
+DEAD_POLICE_TABLE_TEMPLATE = """
+<p xmlns="">
+      Killed in the line of duty:
+   </p>
+<table xmlns="" class="playerlist">
+  <tr><th>Rank</th><th>Pseudonym</th><th>Real Name</th><th>Email Address</th><th>College</th><th>Notes</th></tr>
+  {ROWS}
+</table>
+"""
+
+DEAD_POLICE_TABLE_ROW_TEMPLATE = """
+<tr><td>{RANK}</td><td>{PSEUDONYM}</td><td>{NAME}</td><td>{EMAIL}</td><td>{COLLEGE}</td><td>{NOTES}</td></tr>
+"""
+
+DEAD_CORRUPT_POLICE_TABLE_TEMPLATE = """
+<p xmlns="">
+      And these are the names of those who turned to the dark side, and paid the price:
+   </p>
+<table xmlns="" class="playerlist">
+  <tr><th>Rank</th><th>Pseudonym</th><th>Real Name</th><th>Email Address</th><th>College</th><th>Notes</th></tr>
+  {ROWS}
+</table>
+"""
+
+DEAD_CORRUPT_POLICE_TABLE_ROW_TEMPLATE = """
+<tr><td>{RANK}</td><td>{PSEUDONYM}</td><td>{NAME}</td><td>{EMAIL}</td><td>{COLLEGE}</td><td>{NOTES}</td></tr>
+"""
+
+NO_DEAD_POLICE = """<p xmlns="">No police have been killed... yet.</p>"""
+
+POLICE_PAGE_TEMPLATE: str
+with open(os.path.join(ROOT_DIR, "plugins", "custom_plugins", "html_templates", "police.html"), "r") as F:
+    POLICE_PAGE_TEMPLATE = F.read()
+
+@registered_plugin
+class PolicePlugin(AbstractPlugin):
+    def __init__(self):
+        super().__init__("PolicePlugin")
+
+        self.html_ids = {
+            "Rank": self.identifier + "_rank",
+            "Rank Name": self.identifier + "_name"
+        }
+
+        self.config_exports = [
+            ConfigExport(
+                "PolicePlugin_set_ranks",
+                "Police -> Set Ranks",
+                self.ask_set_ranks,
+                self.answer_set_ranks
+            )
+        ]
+
+        self.ranks = GENERIC_STATE_DATABASE.arb_state.setdefault(self.identifier, DEFAULT_RANKS)
```

Comment:
I think you intend to avoid having to `.setdefault` throughout code with this, but I'm not sure it will work as you intend. If the user nukes the database and then enables the `PolicePlugin`, the `__init__` doesn't get re-run and `GENERIC_STATE_DATABASE["PolicePlugin"]` won't exist (`__init__` is only run on program start)

Reply:
I did encounter this a few times while testing; I assume it can be fixed by using this setdefault everywhere the database is read (in which case do I even need it in __init__()?)

Reply:
It's harmless in `__init__`. I'd leave it there. But yes, the first access to the state in any function must use `.setdefault`.